### PR TITLE
monitor support thanos sidecar feature

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve? 
#3553 

### What is changed and how does it work?
Support thanos sidecar.
### Check List

Tests 

- Unit test
- E2E test
- Stability test
- Manual test 

Code changes

- Has Go code change

Related changes

- Need to cherry-pick to the release branch
- Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support thanos sidecar and can use thanos to do multiple cluster monitoring and management.
```

### CODE REVIEW

**Comment by mikechengwei** [comment]
Thanos sidecar Manual test log:
image

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 1]
```diff
@@ -92,6 +94,13 @@ type TidbMonitorSpec struct {
 
 	// ClusterScoped indicates whether this monitor should manage Kubernetes cluster-wide TiDB clusters
 	ClusterScoped bool `json:"clusterScoped,omitempty"`
+	// The labels to add to any time series or alerts when communicating with
+	// external systems (federation, remote storage, Alertmanager).
+	ExternalLabels map[string]string `json:"externalLabels,omitempty"`
+	// Name of Prometheus external label used to denote replica name.
+	// Defaults to the value of `prometheus_replica`. External label will
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:101]: We should add a comment for the two fields that they are only valid when `spec.thanos` is configured, right?

File: pkg/monitor/monitor/util.go [hunk 2]
```diff
@@ -476,6 +483,9 @@ func getMonitorPrometheusContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.T
 	if monitor.Spec.Prometheus.ImagePullPolicy != nil {
 		c.ImagePullPolicy = *monitor.Spec.Prometheus.ImagePullPolicy
 	}
+	if monitor.Spec.Thanos != nil {
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:486]: Move this to L468 where the command is contructed?

File: pkg/monitor/monitor/util.go [hunk 3]
```diff
@@ -757,6 +767,20 @@ func getMonitorService(monitor *v1alpha1.TidbMonitor) []*core.Service {
 			prometheusService.Spec.LoadBalancerSourceRanges = monitor.Spec.Prometheus.Service.LoadBalancerSourceRanges
 		}
 	}
+
+	if monitor.Spec.Thanos != nil {
+		prometheusService.Spec.Ports = append(prometheusService.Spec.Ports, core.ServicePort{
+			Name:       "thanossidecar-grpc",
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:773]:
```suggestion
			Name:       "thanos-grpc",
```

File: pkg/monitor/monitor/util.go [hunk 4]
```diff
@@ -757,6 +767,20 @@ func getMonitorService(monitor *v1alpha1.TidbMonitor) []*core.Service {
...
+			Protocol:   core.ProtocolTCP,
+			Port:       10901,
+			TargetPort: intstr.FromInt(10901),
+		}, core.ServicePort{
+			Name:       "thanossidecar-http",
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:778]:
```suggestion
			Name:       "thanos-http",
```

File: pkg/monitor/monitor/util.go [hunk 5]
```diff
@@ -1003,3 +1032,124 @@ func getMonitorVolumeClaims(monitor *v1alpha1.TidbMonitor) []core.PersistentVolu
...
+				Protocol:      "TCP",
+			},
+			{
+				Name:          "grpc",
+				ContainerPort: 10901,
+				Protocol:      "TCP",
+			},
+		},
+	}
+	if monitor.Spec.Thanos.ObjectStorageConfig != nil || monitor.Spec.Thanos.ObjectStorageConfigFile != nil {
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:1087]: The code block here does not match with the below comment:
```
// ObjectStorageConfigFile specifies the path of the object storage configuration file.
	// When used alongside with ObjectStorageConfig, ObjectStorageConfigFile takes precedence.
```

File: pkg/monitor/monitor/util.go [hunk 6]
```diff
@@ -1003,3 +1032,124 @@ func getMonitorVolumeClaims(monitor *v1alpha1.TidbMonitor) []core.PersistentVolu
...
+			container.Args = append(container.Args, "--objstore.config-file="+*monitor.Spec.Thanos.ObjectStorageConfigFile)
+		}
+		storageDir := "/prometheus"
+		container.Args = append(container.Args, fmt.Sprintf("--tsdb.path=%s", storageDir))
+		container.VolumeMounts = append(
+			container.VolumeMounts,
+			core.VolumeMount{
+				Name:      v1alpha1.TidbMonitorMemberType.String(),
+				MountPath: storageDir,
+				SubPath:   "prometheus-db",
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:1108]: What's this used for? It seems that it's not used.

File: pkg/monitor/monitor/util.go [hunk 7]
```diff
@@ -1003,3 +1032,124 @@ func getMonitorVolumeClaims(monitor *v1alpha1.TidbMonitor) []core.PersistentVolu
 	}
 	return nil
 }
+
+func getThanosSidecarContainer(monitor *v1alpha1.TidbMonitor) core.Container {
+	bindAddress := "[$(POD_IP)]"
+	if monitor.Spec.Thanos.ListenLocal {
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:1038]: What about defining a variable `thanos:=monitor.Spec.Thanos` to simplify the code.

File: pkg/monitor/monitor/util.go [hunk 8]
```diff
@@ -1003,3 +1032,124 @@ func getMonitorVolumeClaims(monitor *v1alpha1.TidbMonitor) []core.PersistentVolu
...
+	replicaExternalLabelName := defaultReplicaExternalLabelName
+	if monitor.Spec.ReplicaExternalLabelName != nil {
+		if *monitor.Spec.ReplicaExternalLabelName != "" {
+			replicaExternalLabelName = *monitor.Spec.ReplicaExternalLabelName
+		} else {
+			replicaExternalLabelName = ""
+		}
+	}
+	if replicaExternalLabelName != "" {
+		m[model.LabelName(replicaExternalLabelName)] = "$(POD_NAME)"
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:1149]: Where is `$(POD_NAME)` defined?

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 9]
```diff
@@ -159,6 +168,35 @@ type InitializerSpec struct {
 	Envs map[string]string `json:"envs,omitempty"`
 }
 
+// InitializerSpec is the desired state of initializer
+type ThanosSpec struct {
+	MonitorContainer `json:",inline"`
+	// ObjectStorageConfig configures object storage in Thanos.
+	// Alternative to ObjectStorageConfigFile, and lower order priority.
+	ObjectStorageConfig *corev1.SecretKeySelector `json:"objectStorageConfig,omitempty"`
```
[the remark this answers is not in the document]
Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:176]: Reference https://thanos.io/tip/thanos/storage.md/#configuration .

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:101]: The labels to add to any time series or alerts when communicating with
external systems (federation, remote storage, Alertmanager).

[the remark this answers is not in the document]
Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:101]: The external label is the configuration of prometheus, so it is not an exclusive function of thanos

File: pkg/monitor/monitor/util.go [hunk 10]
```diff
@@ -1003,3 +1033,131 @@ func getMonitorVolumeClaims(monitor *v1alpha1.TidbMonitor) []core.PersistentVolu
...
+				Name: "OBJSTORE_CONFIG",
+				ValueFrom: &core.EnvVarSource{
+					SecretKeyRef: thanos.ObjectStorageConfig,
+				},
+			})
+		} else {
+			if thanos.ObjectStorageConfigFile != nil {
+				container.Args = append(container.Args, "--objstore.config-file="+*thanos.ObjectStorageConfigFile)
+			}
+		}
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:1109]:
```suggestion
        if thanos.ObjectStorageConfigFile != nil {
				container.Args = append(container.Args, "--objstore.config-file="+*thanos.ObjectStorageConfigFile)
		} else {
			container.Args = append(container.Args, "--objstore.config=$(OBJSTORE_CONFIG)")
			container.Env = append(container.Env, core.EnvVar{
				Name: "OBJSTORE_CONFIG",
				ValueFrom: &core.EnvVarSource{
					SecretKeyRef: thanos.ObjectStorageConfig,
				},
			})
		}
```

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 11]
```diff
@@ -212,3 +250,14 @@ type DeploymentStorageStatus struct {
...
+
+// TLSConfig extends the safe TLS configuration with file parameters.
+// +k8s:openapi-gen=true
+type TLSConfig struct {
+	// Path to the CA cert in the Prometheus container to use for the targets.
+	CAFile string `json:"caFile,omitempty"`
+	// Path to the client cert file in the Prometheus container for the targets.
+	CertFile string `json:"certFile,omitempty"`
+	// Path to the client key file in the Prometheus container for the targets.
+	KeyFile string `json:"keyFile,omitempty"`
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:262]: How about using a secret name here? as we did for other TLS configs.

Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:262]: The current prometheus server does not support tls function. What about tls that we don't support in this pr?

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 12]
```diff
@@ -159,6 +168,35 @@ type InitializerSpec struct {
...
+	// Alternative to ObjectStorageConfigFile, and lower order priority.
+	ObjectStorageConfig *corev1.SecretKeySelector `json:"objectStorageConfig,omitempty"`
+	// ObjectStorageConfigFile specifies the path of the object storage configuration file.
+	// When used alongside with ObjectStorageConfig, ObjectStorageConfigFile takes precedence.
+	ObjectStorageConfigFile *string `json:"objectStorageConfigFile,omitempty"`
+	// ListenLocal makes the Thanos sidecar listen on loopback, so that it
+	// does not bind against the Pod IP.
+	ListenLocal bool `json:"listenLocal,omitempty"`
+	// TracingConfig configures tracing in Thanos. This is an experimental feature, it may change in any upcoming release in a breaking way.
+	TracingConfig *corev1.SecretKeySelector `json:"tracingConfig,omitempty"`
```
[the remark this answers is not in the document]
Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:184]: Do you mean add --tracing.config-file  configuration ?

[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:184]: YES

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go (hunk 11 above)
Comment (csuzhangxc) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:262]: SGTM

File: pkg/monitor/monitor/util.go [hunk 13]
```diff
@@ -458,6 +465,9 @@ func getMonitorPrometheusContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.T
 	if monitor.Spec.Prometheus.Config != nil && len(monitor.Spec.Prometheus.Config.CommandOptions) > 0 {
 		c.Command = append(c.Command, monitor.Spec.Prometheus.Config.CommandOptions...)
 	}
+	if monitor.Spec.Thanos != nil {
+		c.Command = append(c.Command, "--storage.tsdb.max-block-duration=2h")
```
Comment (csuzhangxc) [inline pkg/monitor/monitor/util.go:469]: Do we need to set `--storage.tsdb.min-block-duration` here too?

File: pkg/monitor/monitor/util.go [hunk 14]
```diff
@@ -1003,3 +1033,131 @@ func getMonitorVolumeClaims(monitor *v1alpha1.TidbMonitor) []core.PersistentVolu
...
+	replicaExternalLabelName := defaultReplicaExternalLabelName
+	if monitor.Spec.ReplicaExternalLabelName != nil {
+		if *monitor.Spec.ReplicaExternalLabelName != "" {
+			replicaExternalLabelName = *monitor.Spec.ReplicaExternalLabelName
+		} else {
+			replicaExternalLabelName = ""
+		}
+	}
+	if replicaExternalLabelName != "" {
+		m[model.LabelName(replicaExternalLabelName)] = "$(POD_NAME)"
```
Comment (csuzhangxc) [inline pkg/monitor/monitor/util.go:1157]: It is needed to append other parts as the label value? like namespace?

**Comment by csuzhangxc** [comment]
Will you add some tests to query metrics from object storage in later PRs?

**Comment by mikechengwei** [comment]
> Will you add some tests to query metrics from object storage in later PRs?

Do you mean e2e test ?

File: pkg/monitor/monitor/util.go (hunk 14 above)
Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:1157]: Does adding namespace guarantee uniqueness?

File: pkg/monitor/monitor/util.go (hunk 13 above)
Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:469]: I think we need.

File: pkg/monitor/monitor/util.go (hunk 14 above)
Comment (csuzhangxc) [reply pkg/monitor/monitor/util.go:1155]: yes, it seems this label value should be unique.

**Comment by csuzhangxc** [comment]
> > Will you add some tests to query metrics from object storage in later PRs?
> 
> Do you mean e2e test ?

YES

File: pkg/monitor/monitor/util.go (hunk 14 above)
Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:1155]: I don't think the Ns tag is needed. If this is to add ns to the environment variable, we just provide a solution. Users can also customize these labels.

File: pkg/monitor/monitor/util.go (hunk 13 above)
Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:469]: When we use thanos ,we need to disable prometheus compact.The --storage.tsdb.min-block-duration and --storage.tsdb.max-block-duration must be set to equal values to disable local compaction .

Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:469]: https://thanos.io/v0.11/components/sidecar/

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 15]
```diff
@@ -108,6 +117,9 @@ type PrometheusSpec struct {
 
 	// +optional
 	Config *PrometheusConfiguration `json:"config,omitempty"`
+
+	// Disable prometheus compaction.
+	DisableCompaction bool `json:"disableCompaction,omitempty"`
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:122]: How about always disable compaction if `monitor.Spec.Thanos != nil` as you did before?

File: pkg/monitor/monitor/util.go (hunk 14 above)
Comment (csuzhangxc) [reply pkg/monitor/monitor/util.go:1160]: @DanielZhangQD WDYT

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go (hunk 15 above)
Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:122]: Done

File: pkg/monitor/monitor/util.go (hunk 14 above)
Comment (DanielZhangQD) [reply pkg/monitor/monitor/util.go:1160]: I think it's OK to add the namespace to the value.

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**